src/gtest/test_upgrades.cpp

```diff
@@ -0,0 +1,147 @@
+    EXPECT_EQ(CurrentEpochBranchId(1000000, params), 0);
+
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY, Consensus::NetworkUpgrade::ALWAYS_ACTIVE);
+
+    EXPECT_EQ(CurrentEpoch(0, params), Consensus::UPGRADE_TESTDUMMY);
+    EXPECT_EQ(CurrentEpochBranchId(0, params), nBranchId);
+    EXPECT_EQ(CurrentEpoch(1000000, params), Consensus::UPGRADE_TESTDUMMY);
+    EXPECT_EQ(CurrentEpochBranchId(1000000, params), nBranchId);
+
+    int nActivationHeight = 100;
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY, nActivationHeight);
+
+    EXPECT_EQ(CurrentEpoch(0, params), Consensus::UPGRADE_SPROUT);
+    EXPECT_EQ(CurrentEpochBranchId(0, params), 0);
+    EXPECT_EQ(CurrentEpoch(nActivationHeight - 1, params), Consensus::UPGRADE_SPROUT);
+    EXPECT_EQ(CurrentEpochBranchId(nActivationHeight - 1, params), 0);
+    EXPECT_EQ(CurrentEpoch(nActivationHeight, params), Consensus::UPGRADE_TESTDUMMY);
+    EXPECT_EQ(CurrentEpochBranchId(nActivationHeight, params), nBranchId);
+    EXPECT_EQ(CurrentEpoch(1000000, params), Consensus::UPGRADE_TESTDUMMY);
+    EXPECT_EQ(CurrentEpochBranchId(1000000, params), nBranchId);
+
+    // Revert to default
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY, Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT);
+}
+
+TEST(Upgrades, IsActivationHeight) {
+    SelectParams(CBaseChainParams::REGTEST);
+    const Consensus::Params& params = Params().GetConsensus();
+
+    // Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(-1, params));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(0, params));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(1, params));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(1000000, params));
+
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY, Consensus::NetworkUpgrade::ALWAYS_ACTIVE);
+
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(-1, params));
+    EXPECT_TRUE(IsActivationHeightForAnyUpgrade(0, params));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(1, params));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(1000000, params));
+
+    int nActivationHeight = 100;
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_TESTDUMMY, nActivationHeight);
+
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(-1, params));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(0, params));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(1, params));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(nActivationHeight - 1, params));
+    EXPECT_FALSE(IsActivationHeight(nActivationHeight - 1, params, Consensus::UPGRADE_TESTDUMMY));
+    EXPECT_TRUE(IsActivationHeightForAnyUpgrade(nActivationHeight, params));
+    EXPECT_TRUE(IsActivationHeight(nActivationHeight, params, Consensus::UPGRADE_TESTDUMMY));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(nActivationHeight + 1, params));
+    EXPECT_FALSE(IsActivationHeightForAnyUpgrade(1000000, params));
```

Comment:
Add test coverage for ``IsActivationHeight(..., Consensus::UPGRADE_TESTDUMMY)`` with heights ``-1``, ``0``, ``1``, ``nActivationHeight + 1``, ``1000000``.